File: src/backend/decoders/pop3.py
def decodePOP3(rawPayload):
    """
    Decode POP3 commands and responses from raw payload bytes.
    Returns a dict with Type (Command/Response), command/status, and argument/message,
    or None if the payload is not recognisable as POP3 traffic.
    """
    POP3_COMMANDS = {
        "USER",
        "PASS",
        "APOP",
        "QUIT",
        "STAT",
        "LIST",
        "RETR",
        "DELE",
        "NOOP",
        "RSET",
        "TOP",
        "UIDL",
        "CAPA",
        "AUTH",
        "STLS",
    }
    try:
        text = rawPayload.decode(errors="ignore")
        lines = text.replace("\r\n", "\n").split("\n")
        firstLine = lines[0].strip()
        if not firstLine:
            return None
        parts = firstLine.split(" ", 1)
        word = parts[0].upper()
        if word in POP3_COMMANDS:
            arg = parts[1].strip() if len(parts) > 1 else ""
            result = {
                "Type": "Command",
                "pop3.type": "Command",
                "Command": word,
                "pop3.command": word,
                "Argument": arg,
                "pop3.argument": arg,
            }
            if word == "USER" and arg:
                result["Credentials"] = {"username": arg}
            elif word == "PASS" and arg:
                result["Credentials"] = {"password": arg}
                result["Argument"] = "***"
                result["pop3.argument"] = "***"
            return result
        if word in ("+OK", "-ERR"):
            message = parts[1].strip() if len(parts) > 1 else ""
            return {
                "Type": "Response",
                "pop3.type": "Response",
                "Status": word,
                "pop3.status": word,
                "Message": message,
                "pop3.message": message,
            }
        return None
    except Exception:
        return None

File: src/ui/decoders/conv/pop3.js
// POP3 Conv decoder: parses +OK / -ERR responses and recognized POP3 commands.

const POP3_COMMANDS = new Set([
    "USER",
    "PASS",
    "STAT",
    "LIST",
    "RETR",
    "DELE",
    "NOOP",
    "RSET",
    "QUIT",
    "APOP",
    "TOP",
    "UIDL",
]);

function decodePop3FromBytes(bytes) {
    const text = new TextDecoder("utf-8", { fatal: false }).decode(bytes);
    const lines = text.split(/\r?\n/).filter((l) => l.trim());
    if (!lines.length) return null;
    const fields = [];
    let detected = false;
    for (const line of lines) {
        if (line.startsWith("+OK")) {
            fields.push({ name: "Response", value: "+OK" });
            const msg = line.slice(3).trim();
            if (msg) fields.push({ name: "Message", value: msg });
            detected = true;
        } else if (line.startsWith("-ERR")) {
            fields.push({ name: "Response", value: "-ERR" });
            const msg = line.slice(4).trim();
            if (msg) fields.push({ name: "Error", value: msg });
            detected = true;
        } else {
            const parts = line.split(/\s+/);
            const cmd = parts[0].toUpperCase();
            if (POP3_COMMANDS.has(cmd)) {
                fields.push({ name: "Command", value: cmd });
                if (parts.length > 1) {
                    fields.push({ name: "Argument", value: parts.slice(1).join(" ") });
                }
                detected = true;
            }
        }
    }
    if (!detected) return null;
    return { protocol: "POP3", fields };
}

module.exports = { decodePop3FromBytes };
